### PR TITLE
chore: Add spilling metrics of SortMergeJoin

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #.

## Rationale for this change

We don't propagate spilling metrics of SortMergeJoin from DataFusion to Comet yet. This patch adds spilling related metrics to Comet SortMergeJoin operator.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/shuffle/sort/RowPartition.scala [hunk 1]
```diff
@@ -30,7 +30,11 @@ class RowPartition(initialSize: Int) {
     rowSizes += size
   }
 
-  def getNumRows: Int = rowAddresses.size
+  def getNumRows: Int = if (rowAddresses == null) {
+    0
+  } else {
+    rowAddresses.size
+  }
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/shuffle/sort/RowPartition.scala:37]: This is unrelated. Just when I tested several memory configs to try to trigger spilling, this method throws NPE sometimes.

File: spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala [hunk 2]
```diff
@@ -519,6 +519,8 @@ class CometExecSuite extends CometTestBase {
           assert(metrics("peak_mem_used").value > 1L)
           assert(metrics.contains("join_time"))
           assert(metrics("join_time").value > 1L)
+          assert(metrics.contains("spill_count"))
+          assert(metrics("spill_count").value == 0)
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala:523]: Spilling is not triggered. But it should propagate related metrics.

Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala:523]: Can we also have a test that forces spilling?

**Review by comphead — APPROVED** [review]
lgtm thanks @viirya just wondering if those numbers taken from Spark context or from DF spilled metrics?

**Comment by viirya** [comment]
> lgtm thanks @viirya just wondering if those numbers taken from Spark context or from DF spilled metrics?

They are from DataFusion SortMergeJoin operator's metrics. Comet operators' metrics are propagated from DataFusion.

File: spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala (hunk 2 above)
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala:523]: I will keep trying but wanted to merge these metrics first.

**Comment by viirya** [comment]
Merged. Thanks @andygrove @comphead
